central/complianceoperator/v2/scanconfigurations/service/download.go

```diff
@@ -0,0 +1,138 @@
+		return
+	}
+
+	snapshot, found, err := h.snapshotDataStore.GetSnapshot(ctx, id)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Error retrieving snapshot with the given id %s", id))
+		return
+	}
+	if !found {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Unable to find the snapshot with the given id %s", id))
+		return
+	}
+
+	if slimUser.GetId() != snapshot.GetUser().GetId() {
+		httputil.WriteGRPCStyleError(w, codes.PermissionDenied, errors.Errorf("Report %s cannot be downloaded by the user %s", id, slimUser.GetId()))
+		return
+	}
+
+	status := snapshot.GetReportStatus()
+	if status.GetReportNotificationMethod() != storage.ComplianceOperatorReportStatus_DOWNLOAD {
+		httputil.WriteGRPCStyleError(w, codes.InvalidArgument, errors.Errorf("Report %s did not generated a downloaded report", id))
+		return
+	}
+
+	switch status.GetRunState() {
+	case storage.ComplianceOperatorReportStatus_FAILURE:
+		httputil.WriteGRPCStyleError(w, codes.FailedPrecondition, errors.Errorf("Report %s failed", id))
+		return
+	case storage.ComplianceOperatorReportStatus_WAITING, storage.ComplianceOperatorReportStatus_PREPARING:
+		httputil.WriteGRPCStyleError(w, codes.Unavailable, errors.Errorf("Report %s is not ready for download", id))
+		return
+	}
+
+	buf := bytes.NewBuffer(nil)
+	ctx = sac.WithGlobalAccessScopeChecker(ctx,
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Administration),
+		),
+	)
+
+	_, exists, err := h.blobStore.Get(ctx, common.GetComplianceReportBlobPath(snapshot.GetScanConfigurationId(), id), buf)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.Internal, errors.New("Failed to fetch report data"))
+		return
+	}
+	if !exists {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Report %s is not available to download", id))
+		return
+	}
+
+	// Tell the browser this is a download.
+	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="report-%s.zip"`, zip.GetSafeFilename(id)))
+	w.Header().Set("Content-Type", "application/zip")
+	w.Header().Set("Content-Length", fmt.Sprint(buf.Len()))
+	_, err = w.Write(buf.Bytes())
+
+	writeSnapshotCtx := sac.WithGlobalAccessScopeChecker(ctx,
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Compliance)),
+	)
+	if err == nil && status.GetRunState() == storage.ComplianceOperatorReportStatus_GENERATED {
+		snapshot.GetReportStatus().RunState = storage.ComplianceOperatorReportStatus_DELIVERED
+		err = h.snapshotDataStore.UpsertSnapshot(writeSnapshotCtx, snapshot)
+		if err != nil {
+			log.Error("Error setting report state to DELIVERED")
+		}
+	}
+}
+
+func parseReportID(r *http.Request) (string, error) {
+	err := r.ParseForm()
+	if err != nil {
+		return "", err
```

Comment:
When we return an error here, then we don't have a context information that this is about parsing the ID. It is not a big issue, but maybe we could do wrapping in the place where we call `parseReportID`? (That may alse require to tune the wording of the "empty report id", to avoid repetition).

Reply:
Is this all validation that we want to do? Shouldn't we check for more - for example, that the ID contains only letters and digits and dashes?

Reply:
I've returned a different error in the `ParseForm` path but I think that's not quite what you were saying. I think wrapping the error in the caller is fine but I don't know what to return in case the id is empty to avoid repetition. Maybe we can wrap the error with `invalid request` and keep the `empty report id`?

> Is this all validation that we want to do? Shouldn't we check for more - for example, that the ID contains only letters and digits and dashes?

I'm not sure about this. We could expose the type of id that the backend is expecting unnecessarily. If the user calls with an invalid id the db will return nothing.

Reply:
This is a nit and is not crucial to be improved, but I was thinking about something like this:

```go
func parseReportID(r *http.Request) (string, error) {
	err := r.ParseForm()
	if err != nil {
		return "", errors.Wrap(err, "failed to parse the request form")
	}
	var id string
	if id = r.Form.Get("id"); id == "" {
		return "", errors.New("ID is empty")
	}
	return id, nil
}

// ...

id, err := parseReportID(r)
	if err != nil {
		httputil.WriteGRPCStyleError(w, codes.InvalidArgument, errors.Wrapf(err, "error parsing report ID"))
		return
	}
```

That would produce the following error strings:
1. `InvalidArgument: error parsing report ID: failed to parse request form: <whatever reason was there>`
2. `InvalidArgument: error parsing report ID: ID is empty`

Reply:
Done